### PR TITLE
Modify EMP creator and collateral whitelist for UMIPs 8 and 9

### PR DESCRIPTION
**Pull request by mrice32**
This PR implements the smart contract changes needed for UMIPs 8 and 9.

Changes:
- The collateral currency whitelist normally associated with a particular EMP creator is now tracked by the finder.
- The EMP creator no longer takes the whitelist in on construction -- it looks the whitelist up via the finder.
- The EMP creator now allows any future timestamp to be taken in as the expiry.
- The EMP creator now allows liquidation and withdrawal livenesses to be passed in.

These changes involve abi changes to the EMP creator contract. So I think we should probably create a repository tag before merging this to stamp the version (this is not a practice we've employed in the past, but since we aren't to the point of uploading npm packages just yet, a repo tag is probably a lightweight way to mark the previous abi.

I think the test modifications should cover all the changed functionality, but please comment if you think any additional testing should be done.

### CODE REVIEW

File: core/contracts/financial-templates/expiring-multiparty/ExpiringMultiPartyCreator.sol [hunk 1]
```diff
@@ -36,69 +36,26 @@ contract ExpiringMultiPartyCreator is ContractCreator, Testable, Lockable {
...
      * @param _tokenFactoryAddress ERC20 token factory used to deploy synthetic token instances.
      * @param _timerAddress Contract that stores the current time in a testing environment.
      */
     constructor(
         address _finderAddress,
-        address _collateralTokenWhitelist,
         address _tokenFactoryAddress,
         address _timerAddress
     ) public ContractCreator(_finderAddress) Testable(_timerAddress) nonReentrant() {
-        collateralTokenWhitelist = AddressWhitelist(_collateralTokenWhitelist);
```
Comment (nicholaspai) [inline core/contracts/financial-templates/expiring-multiparty/ExpiringMultiPartyCreator.sol:79]: Do we still need to import AddressWhitelist now?

File: core/contracts/oracle/implementation/ContractCreator.sol [hunk 2]
```diff
@@ -15,6 +16,14 @@ abstract contract ContractCreator {
         finderAddress = _finderAddress;
     }
 
+    function _requireWhitelistedCollateral(address collateralAddress) internal view {
```
Comment (nicholaspai) [inline core/contracts/oracle/implementation/ContractCreator.sol:19]: +1 I like where you placed the logical check for whether a currency is whitelisted. Any creator should check with the CollateralCurrencyWhitelist via this internal call.

File: core/migrations/14_deploy_expiring_multi_party_creator.js [hunk 3]
```diff
@@ -8,21 +8,32 @@ const Timer = artifacts.require("Timer");
...
 const TestnetERC20 = artifacts.require("TestnetERC20");
 const { RegistryRolesEnum } = require("../../common/Enums.js");
+const { interfaceName } = require("../utils/Constants.js");
 
 module.exports = async function(deployer, network, accounts) {
   const keys = getKeysForNetwork(network, accounts);
   const controllableTiming = enableControllableTiming(network);
 
-  // Deploy whitelists.
-  const { contract: collateralCurrencyWhitelist } = await deploy(deployer, network, AddressWhitelist, {
```
Comment (nicholaspai) [inline core/migrations/14_deploy_expiring_multi_party_creator.js:17]: So we can remove the `AddressWhitelist` import from this script right?

File: core/migrations/15_deploy_local_weth.js [hunk 4]
```diff
@@ -25,8 +24,6 @@ module.exports = async function(deployer, network, accounts) {
   }
 
   // Add wethTokenAddress to the margin currency whitelist.
-  const empCreator = await ExpiringMultiPartyCreator.deployed();
-  const collateralWhitelistAddress = await empCreator.collateralTokenWhitelist();
-  const collateralWhitelist = await AddressWhitelist.at(collateralWhitelistAddress);
+  const collateralWhitelist = await AddressWhitelist.deployed();
```
Comment (nicholaspai) [inline core/migrations/15_deploy_local_weth.js:27]: Shouldn't this be CollateralWhitelist.deployed()

**Comment by mrice32** [comment]
> Looks great. Please review my nit comments.
> 
> One high level comment: Do we need the AddressWhitelist anymore or can we more deprecate it completely from the contracts repo?

Thanks for the quick review! Will respond to your comments tomorrow.

Yeah, this is actually a little confusing and something I want comments on -- `AddressWhitelist` is the name of the general _contract_ that's used to manage a whitelist of addresses, but collateral whitelist is how I'm referring to this _usage_ of that contract. This is how we were doing it previously because we used to use `AddressWhitelist` for other address whitelists -- now I think it's only used for this.

Thoughts on whether we should just change the name of the contract to `CollateralWhitelist`? Or we could create a something like the EMP-Liquidatable relationship, where we have `CollateralWhitelist` inherit from `AddressWhitelist`, but have it make no changes to the implementation -- just so it has a distinct name.

File: core/contracts/financial-templates/expiring-multiparty/ExpiringMultiPartyCreator.sol (hunk 1 above)
Comment (nicholaspai) [reply core/contracts/financial-templates/expiring-multiparty/ExpiringMultiPartyCreator.sol:79]: Ah wow I totally misread the fact that `collateralTokenWhitelist` is just an instance of `AddressWhitelist`. Please ignore all my comments about this.

**Comment by nicholaspai** [comment]
> > Looks great. Please review my nit comments.
> > One high level comment: Do we need the AddressWhitelist anymore or can we more deprecate it completely from the contracts repo?
> 
> Thanks for the quick review! Will respond to your comments tomorrow.
> 
> Yeah, this is actually a little confusing and something I want comments on -- `AddressWhitelist` is the name of the general _contract_ that's used to manage a whitelist of addresses, but collateral whitelist is how I'm referring to this _usage_ of that contract. This is how we were doing it previously because we used to use `AddressWhitelist` for other address whitelists -- now I think it's only used for this.
> 
> Thoughts on whether we should just change the name of the contract to `CollateralWhitelist`? Or we could create a something like the EMP-Liquidatable relationship, where we have `CollateralWhitelist` inherit from `AddressWhitelist`, but have it make no changes to the implementation -- just so it has a distinct name.


I completely misread this. Please ignore. I think `AddressWhitelist` is the correct name of the contract because its literally just a whitelist for addresses. I think you are right in making the instance name more specific about **what** sort of addresses its whitelisting: collateral currencies.

File: core/contracts/financial-templates/expiring-multiparty/ExpiringMultiPartyCreator.sol [hunk 5]
```diff
@@ -155,5 +104,7 @@ contract ExpiringMultiPartyCreator is ContractCreator, Testable, Lockable {
         constructorParams.sponsorDisputeRewardPct = params.sponsorDisputeRewardPct;
         constructorParams.disputerDisputeRewardPct = params.disputerDisputeRewardPct;
         constructorParams.minSponsorTokens = params.minSponsorTokens;
+        constructorParams.withdrawalLiveness = params.withdrawalLiveness;
```
Comment (chrismaree) [inline core/contracts/financial-templates/expiring-multiparty/ExpiringMultiPartyCreator.sol:107]: With this change, there is no validation done on either of these params at all; not even a check to ensure that theses values are `!=0`. This is fine I guess but would we not want to impose _some_ kind of bounds on these?

Comment (mrice32) [reply core/contracts/financial-templates/expiring-multiparty/ExpiringMultiPartyCreator.sol:107]: No, you’re 100% correct. We should validate that they’re > 0. That’s an oversight. Will fix!

Do you think a >0 check is sufficient — just to make sure the liveness values force the caller to wait at least 1 block?

**Comment by mrice32** [comment]
@chrismaree 

> 1. we now no longer impose strict timestamps. this means two things: a) contracts can settle at arbitrary times, rather than at "nice" date-time stamps. b) there is no longer a hard "stop" on when the contract will still operate.

Yeah, it seems like what you're getting at is that someone _could_ set the expiry to 500 yrs from now, and then the contract would a) have very little reason to trade near its underlier and b) mean that this version would need to be "supported" indefinitely by the DVM, etc. I think both of these are okay. For a, it's not outside the realm of reality that someone were to use the EMP to create a perp token if the identifier were constructed in a certain way -- so I see no reason to put a hard cap on this parameter as a high expiry _could_ make sense. For b, I think the DVM should utilize the emergency shutdown (with plenty of pre-warning) to deprecate contract/abi versions that it needs to cease supporting rather than relying on hard-caps on expiries, but that's just my weakly held opinion (deprecation is always pretty messy).

> 2. as we no longer have any `STRICT_LIQUIDATION_LIVENESS` or `STRICT_WITHDRAWAL_LIVENESS` there is no longer any validation on these liveness parameters. I could, in theory, set these to 0 and deploy contracts that would be completely broken.

100% agree. Setting these to 0 completely breaks the incentives since you could literally execute a liquidation or withdraw request _and_ the withdrawal in the same txn or block. I don't really see any possible, legitimate use-case for using the contract in that way. I think the most sensible check is just that they're > 0, meaning they couldn't be executed in the same block. IMO any larger lower bound would be pretty arbitrary and probably wouldn't provide much protection, anyway.

Thoughts?

File: core/contracts/financial-templates/expiring-multiparty/ExpiringMultiPartyCreator.sol (hunk 5 above)
Comment (mrice32) [reply core/contracts/financial-templates/expiring-multiparty/ExpiringMultiPartyCreator.sol:107]: I'll make this change.

Comment (mrice32) [reply core/contracts/financial-templates/expiring-multiparty/ExpiringMultiPartyCreator.sol:109]: Done. PTAL.

**Comment by mrice32** [comment]
> > > Looks great. Please review my nit comments.
> > > One high level comment: Do we need the AddressWhitelist anymore or can we more deprecate it completely from the contracts repo?
> > 
> > 
> > Thanks for the quick review! Will respond to your comments tomorrow.
> > Yeah, this is actually a little confusing and something I want comments on -- `AddressWhitelist` is the name of the general _contract_ that's used to manage a whitelist of addresses, but collateral whitelist is how I'm referring to this _usage_ of that contract. This is how we were doing it previously because we used to use `AddressWhitelist` for other address whitelists -- now I think it's only used for this.
> > Thoughts on whether we should just change the name of the contract to `CollateralWhitelist`? Or we could create a something like the EMP-Liquidatable relationship, where we have `CollateralWhitelist` inherit from `AddressWhitelist`, but have it make no changes to the implementation -- just so it has a distinct name.
> 
> I completely misread this. Please ignore. I think `AddressWhitelist` is the correct name of the contract because its literally just a whitelist for addresses. I think you are right in making the instance name more specific about **what** sort of addresses its whitelisting: collateral currencies.

@nicholaspai gotcha. SGTM

File: core/contracts/financial-templates/expiring-multiparty/ExpiringMultiPartyCreator.sol [hunk 6]
```diff
@@ -137,12 +89,11 @@ contract ExpiringMultiPartyCreator is ContractCreator, Testable, Lockable {
         constructorParams.timerAddress = timerAddress;
 
         // Enforce configuration constraints.
-        require(_isValidTimestamp(params.expirationTimestamp), "Invalid expiration timestamp");
         require(bytes(params.syntheticName).length != 0, "Missing synthetic name");
         require(bytes(params.syntheticSymbol).length != 0, "Missing synthetic symbol");
-        constructorParams.withdrawalLiveness = STRICT_WITHDRAWAL_LIVENESS;
-        constructorParams.liquidationLiveness = STRICT_LIQUIDATION_LIVENESS;
-        require(collateralTokenWhitelist.isOnWhitelist(params.collateralAddress), "Collateral is not whitelisted");
+        require(params.withdrawalLiveness != 0, "Withdrawal liveness cannot be 0");
```
Comment (nicholaspai) [inline core/contracts/financial-templates/expiring-multiparty/ExpiringMultiPartyCreator.sol:94]: what if it's negative?

File: core/contracts/financial-templates/expiring-multiparty/ExpiringMultiPartyCreator.sol (hunk 5 above)
Comment (chrismaree) [reply core/contracts/financial-templates/expiring-multiparty/ExpiringMultiPartyCreator.sol:109]: LGTM! `Liquidation liveness cannot be 0` is _exactly_ 32 chars so we right on the limit here

**Comment by mrice32** [comment]
> looks great! nice and simple implementation. Definitely need to version the repo before merging this PR.
> 
> Open question: what (if any) are the audit considerations for this PR? we've changed some pretty fundamental business logic here; would it be a good idea to get OZ to look over this? I know this would effect deadlines so it could perhaps happen in parallel to not push out product launch.

I agree, in general, that this should be audited. However, since it isn't a change to the core DVM, I think the risks are relatively low for anything DVM related (DoS would be the biggest risk). The biggest potential issue is that it somehow breaks the EMPs that are deployed.

Thoughts on optimistically doing the deployment and sending an async request to get them to take a look?

**Comment by chrismaree** [comment]
> > looks great! nice and simple implementation. Definitely need to version the repo before merging this PR.
> > Open question: what (if any) are the audit considerations for this PR? we've changed some pretty fundamental business logic here; would it be a good idea to get OZ to look over this? I know this would effect deadlines so it could perhaps happen in parallel to not push out product launch.
> 
> I agree, in general, that this should be audited. However, since it isn't a change to the core DVM, I think the risks are relatively low for anything DVM related (DoS would be the biggest risk). The biggest potential issue is that it somehow breaks the EMPs that are deployed.
> 
> Thoughts on optimistically doing the deployment and sending an async request to get them to take a look?

For me, an optimistic audit (launch and get audited & same time) is fine as long as the audit comes back quickly. There is some risk but I think this is low given the scope of the changes.

**Pull request merged**